src/CoreWCF.Kafka.Client/src/CoreWCF/ServiceModel/Channels/KafkaOutputChannel.cs

```diff
@@ -0,0 +1,180 @@
+        {
+            return OnOpenAsync().ToApm(callback, state);
+        }
+
+        protected override void OnEndOpen(IAsyncResult result)
+        {
+            result.ToApmEnd();
+        }
+
+        protected override void OnAbort()
+        {
+            OnClose(DefaultCloseTimeout);
+        }
+
+        protected override void OnClose(TimeSpan timeout)
+        {
+            CancellationTokenSource cts = new(timeout);
+            try
+            {
+                _producer.Flush(cts.Token);
+            }
+            finally
+            {
+                cts.Dispose();
+                _producer.Dispose();
+            }
+        }
+
+        private Task OnCloseAsync(TimeSpan timeout)
+        {
+            OnClose(timeout);
+            return Task.CompletedTask;
+        }
+
+        protected override IAsyncResult OnBeginClose(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            return OnCloseAsync(timeout).ToApm(callback, state);
+        }
+
+        protected override void OnEndClose(IAsyncResult result)
+        {
+            result.ToApmEnd();
+        }
+
+        private ArraySegment<byte> EncodeMessage(Message message)
+        {
+            try
+            {
+                RemoteAddress.ApplyTo(message);
+                return _encoder.WriteMessage(message, int.MaxValue, _parent.BufferManager);
+            }
+            finally
+            {
+                message.Close();
+            }
+        }
+
+        private async Task SendAsync(Message message, CancellationToken token)
+        {
+            ArraySegment<byte> messageBuffer = EncodeMessage(message);
+            try
+            {
+                byte[] bytes = new Span<byte>(messageBuffer.Array, messageBuffer.Offset, messageBuffer.Count).ToArray();
+                await _producer.ProduceAsync(_topic, new Message<Null, byte[]> { Value = bytes }, token);
+            }
+            catch (ProduceException<Null, byte[]> produceException)
+            {
+                throw KafkaChannelHelpers.ConvertProduceException(produceException);
+            }
+            finally
+            {
+                _parent.BufferManager.ReturnBuffer(messageBuffer.Array);
+            }
+        }
+
+        public void Send(Message message) => Send(message, DefaultSendTimeout);
+
+        public void Send(Message message, TimeSpan timeout)
+        {
+            using CancellationTokenSource cts = new(timeout);
+            SendAsync(message, cts.Token).GetAwaiter().GetResult();
+        }
+
+        public IAsyncResult BeginSend(Message message, AsyncCallback callback, object state)
+        {
+            return SendAsync(message, CancellationToken.None).ToApm(callback, state);
+        }
+
+        public IAsyncResult BeginSend(Message message, TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            return BeginSend(message, callback, state);
+        }
```

Comment:
You should use the passed in timeout. I think you should change your method signature on SendAsync to take a TimeSpan instead of CancellationToken. The reason is you need to create a CancellationTokenSource for the timeout, but as that creates a background timer, you need to dispose the CTS when the work is done. Passing that around APM methods is a pain and a lot of code. Instead you can create the CTS in SendAsync and dispose it in the finally block.